ome/ome-model
# Add gradle build

Follow up for #86.  This PR no longer has any C++-related conflicts, and is updated to the latest stable gradle release, 6.7.1 and the latest ome-model master version.  This should keep you up-to-date with the possibilities even if you don't want to merge it at this point in time.

--exclude

This PR adds a gradle build system to replace the existing maven build.  CI tests are now using gradle, testing with JDK 8 and 11.  The source and target versions are 1.8.

This should be a drop-in replacement for all of the maven functionality in current use, and it's vastly shorter and simpler than the maven logic.  This is the most complex bioformats-related package, with multiple components, source code generation and documentation generation, so this conversion is as complex as it will get for all of the decoupled components.  It includes all the needed logic for building, testing and making signed releases.

Testing:

Run `./gradlew build|assemble|test|publishToMavenLocal`

To test packaging changes:

```
mvn
./gradlew assemble
pkgdiff ome-xml/target/ome-xml-6.2.2-SNAPSHOT.jar ome-xml/build/libs/ome-xml-6.2.2-SNAPSHOT.jar
(repeat for other jars, and look at the HTML pages for changes; should be cosmetic only--pom.xml/pom.properties not in jar [maven-specific] and some minor MANIFEST.MF differences)
```

Workflow changes:

- Edit *VERSION* to set the version number.  Was originally done to accommodate Java and C++/CMake, now used to accommodate Java and Python setup.py; can be easily moved into the gradle properties if desired.
- Create *~/.gradle/gradle.properties* (equivalent of *~/.m2/settings.xml*) with this content:

```
signing.gnupg.keyName=<keyid>
ossrhUsername=<username>
ossrhPassword=<password>
```

You might also want:

```
python=python3
```

(can be put into *gradle.properties* in ome-model to customise the single project for your system).

- Run `./gradlew publish` to upload to maven central.  Snapshots to snapshots, releases to staging handled automatically.  Or run `./gradlew publishSnapshot` to upload snapshots only, skipping releases, which is useful for CI builds where signing is not possible.
